README.rst
HTTPie: human-friendly CLI HTTP client for the API era
######################################################

HTTPie (pronounced *aitch-tee-tee-pie*) is a command-line HTTP client.
Its goal is to make CLI interaction with web services as human-friendly as possible.
HTTPie is designed for testing, debugging, and generally interacting with APIs & HTTP servers.
The ``http`` & ``https`` commands allow for creating and sending arbitrary HTTP requests.
They use simple and natural syntax and provide formatted and colorized output.



.. class:: no-web no-pdf

    |docs| |pypi| |build| |coverage| |downloads| |gitter|


.. class:: no-web no-pdf

    .. image:: https://raw.githubusercontent.com/httpie/httpie/master/httpie.gif
        :alt: HTTPie in action
        :width: 100%
        :align: center


.. contents::

.. section-numbering::


About this document
===================

This documentation is best viewed at `httpie.org/docs <https://httpie.org/docs>`_.

You can select your corresponding HTTPie version as well as run examples directly from the
browser using a `termible.io <https://termible.io?utm_source=httpie-readme>`_ embedded terminal.

If you are reading this on GitHub, then this text covers the current *development* version.
You are invited to submit fixes and improvements to the the docs by editing
`README.rst <https://github.com/httpie/httpie/blob/master/README.rst>`_.


Main features
=============

* Expressive and intuitive syntax
* Formatted and colorized terminal output
* Built-in JSON support
* Forms and file uploads
* HTTPS, proxies, and authentication
* Arbitrary request data
* Custom headers
* Persistent sessions
* Wget-like downloads
* Linux, macOS and Windows support
* Plugins
* Documentation
* Test coverage


.. class:: no-web

    .. image:: https://raw.githubusercontent.com/httpie/httpie/master/httpie.png
        :alt: HTTPie compared to cURL
        :width: 100%
        :align: center


Installation
============


macOS
-----


On macOS, HTTPie can be installed via `Homebrew <https://brew.sh/>`_
(recommended):

.. code-block:: bash

    $ brew install httpie


A MacPorts *port* is also available:

.. code-block:: bash

    $ port install httpie

Linux
-----

HTTPie is available on the `Snap Store <https://snapcraft.io/httpie>`_:

.. code-block:: bash

    $ snap install httpie

And most Linux distributions provide a package that can be installed using the
system package manager, for example:

.. code-block:: bash

    # Debian, Ubuntu, etc.
    $ apt install httpie

.. code-block:: bash

    # Fedora
    $ dnf install httpie

.. code-block:: bash

    # CentOS, RHEL, ...
    $ yum install httpie

.. code-block:: bash

    # Gentoo
    $ emerge httpie

.. code-block:: bash

    # Arch Linux
    $ pacman -S httpie

.. code-block:: bash

    # Solus
    $ eopkg install httpie


Windows, etc.
-------------

A universal installation method (that works on Windows, macOS, Linux, …,
and always provides the latest version) is to use `pip`_:

.. code-block:: bash

    # Make sure we have an up-to-date version of pip and setuptools:
    $ python -m pip install --upgrade pip setuptools

    $ python -m pip install --upgrade httpie


(If ``pip`` installation fails for some reason, you can try
``easy_install httpie`` as a fallback.)

Windows users can also install HTTPie with `Chocolatey <https://chocolatey.org>`_:

.. code-block:: bash

    $ choco upgrade httpie


Python version
--------------

Python version 3.6 or greater is required.


Unstable version
----------------

You can also install the latest unreleased development version directly from
the ``master`` branch on GitHub.  It is a work-in-progress of a future stable
release so the experience might be not as smooth.


.. class:: no-pdf

|build|


On macOS you can install it with Homebrew:

.. code-block:: bash

    $ brew uninstall --force httpie
    $ brew install --HEAD httpie


Otherwise with ``pip``:

.. code-block:: bash

    $ python -m pip install --upgrade https://github.com/httpie/httpie/archive/master.tar.gz


Verify that now we have the
`current development version identifier <https://github.com/httpie/httpie/blob/master/httpie/__init__.py#L6>`_
with the ``-dev`` suffix, for example:

.. code-block:: bash

    $ http --version
    # 2.0.0-dev


Usage
=====


Hello World:


.. code-block:: bash

    $ https httpie.io/hello


Synopsis:

.. code-block:: bash

    $ http [flags] [METHOD] URL [ITEM [ITEM]]


See also ``http --help``.


Examples
--------

Custom `HTTP method`_, `HTTP headers`_ and `JSON`_ data:

.. code-block:: bash

    $ http PUT pie.dev/put X-API-Token:123 name=John


Submitting `forms`_:

.. code-block:: bash

    $ http -f POST pie.dev/post hello=World


See the request that is being sent using one of the `output options`_:

.. code-block:: bash

    $ http -v pie.dev/get


Build and print a request without sending it using `offline mode`_:

.. code-block:: bash

    $ http --offline pie.dev/post hello=offline


Use `GitHub API`_ to post a comment on an
`issue <https://github.com/httpie/httpie/issues/83>`_
with `authentication`_:

.. code-block:: bash

    $ http -a USERNAME POST https://api.github.com/repos/httpie/httpie/issues/83/comments body='HTTPie is awesome! :heart:'


Upload a file using `redirected input`_:

.. code-block:: bash

    $ http pie.dev/post < files/data.json


Download a file and save it via `redirected output`_:

.. code-block:: bash

    $ http pie.dev/image/png > image.png


Download a file ``wget`` style:

.. code-block:: bash

    $ http --download pie.dev/image/png

Use named `sessions`_ to make certain aspects of the communication persistent
between requests to the same host:


.. code-block:: bash

    $ http --session=logged-in -a username:password pie.dev/get API-Key:123


.. code-block:: bash

    $ http --session=logged-in pie.dev/headers


Set a custom ``Host`` header to work around missing DNS records:

.. code-block:: bash

    $ http localhost:8000 Host:example.com

..


HTTP method
===========

The name of the HTTP method comes right before the URL argument:

.. code-block:: bash

    $ http DELETE pie.dev/delete


Which looks similar to the actual ``Request-Line`` that is sent:

.. code-block:: http

    DELETE /delete HTTP/1.1


When the ``METHOD`` argument is omitted from the command, HTTPie defaults to
either ``GET`` (with no request data) or ``POST`` (with request data).


Request URL
===========

The only information HTTPie needs to perform a request is a URL.

The default scheme is ``http://`` and can be omitted from the argument:

.. code-block:: bash

    $ http example.org
    # => http://example.org


HTTPie also installs an ``https`` executable, where the default
scheme is ``https://``:


.. code-block:: bash

    $ https example.org
    # => https://example.org


Querystring parameters
----------------------

If you find yourself manually constructing URLs with querystring parameters
on the terminal, you may appreciate the ``param==value`` syntax for appending
URL parameters.

With that, you don’t have to worry about escaping the ``&``
separators for your shell. Additionally, any special characters in the
parameter name or value get automatically URL-escaped
(as opposed to parameters specified in the full URL, which HTTPie doesn’t
modify).

.. code-block:: bash

    $ http https://api.github.com/search/repositories q==httpie per_page==1


.. code-block:: http

    GET /search/repositories?q=httpie&per_page=1 HTTP/1.1



URL shortcuts for ``localhost``
-------------------------------

Additionally, curl-like shorthand for localhost is supported.
This means that, for example ``:3000`` would expand to ``http://localhost:3000``
If the port is omitted, then port 80 is assumed.

.. code-block:: bash

    $ http :/foo


.. code-block:: http

    GET /foo HTTP/1.1
    Host: localhost


.. code-block:: bash

    $ http :3000/bar


.. code-block:: http

    GET /bar HTTP/1.1
    Host: localhost:3000


.. code-block:: bash

    $ http :


.. code-block:: http

    GET / HTTP/1.1
    Host: localhost


Other default schemes
---------------------

When HTTPie is invoked as ``https`` then the default scheme is ``https://``
(``$ https example.org`` will make a request to ``https://example.org``).

You can also use the ``--default-scheme <URL_SCHEME>`` option to create
shortcuts for other protocols than HTTP (possibly supported via plugins).
Example for the `httpie-unixsocket <https://github.com/httpie/httpie-unixsocket>`_ plugin:

.. code-block:: bash

    # Before
    $ http http+unix://%2Fvar%2Frun%2Fdocker.sock/info


.. code-block:: bash

    # Create an alias
    $ alias http-unix='http --default-scheme="http+unix"'


.. code-block:: bash

    # Now the scheme can be omitted
    $ http-unix %2Fvar%2Frun%2Fdocker.sock/info


``--path-as-is``
----------------

The standard behaviour of HTTP clients is to normalize the path portion of URLs by squashing dot segments
as a typically filesystem would:


.. code-block:: bash

    $ http -v example.org/./../../etc/password

.. code-block:: http

    GET /etc/password HTTP/1.1


The ``--path-as-is`` option allows you to disable this behavior:

.. code-block:: bash

    $ http --path-as-is -v example.org/./../../etc/password

.. code-block:: http

    GET /../../etc/password HTTP/1.1


Request items
=============

There are a few different *request item* types that provide a
convenient mechanism for specifying HTTP headers, simple JSON and
form data, files, and URL parameters.

They are key/value pairs specified after the URL. All have in
common that they become part of the actual request that is sent and that
their type is distinguished only by the separator used:
``:``, ``=``, ``:=``, ``==``, ``@``, ``=@``, and ``:=@``. The ones with an
``@`` expect a file path as value.

+------------------------------+---------------------------------------------------+
| Item Type                    | Description                                       |
+==============================+===================================================+
| HTTP Headers                 | Arbitrary HTTP header, e.g. ``X-API-Token:123``.  |
| ``Name:Value``               |                                                   |
+------------------------------+---------------------------------------------------+
| URL parameters               | Appends the given name/value pair as a query      |
| ``name==value``              | string parameter to the URL.                      |
|                              | The ``==`` separator is used.                     |
+------------------------------+---------------------------------------------------+
| Data Fields                  | Request data fields to be serialized as a JSON    |
| ``field=value``,             | object (default), to be form-encoded              |
| ``field=@file.txt``          | (with ``--form, -f``), or to be serialized as     |
|                              | ``multipart/form-data`` (with ``--multipart``).   |
+------------------------------+---------------------------------------------------+
| Raw JSON fields              | Useful when sending JSON and one or               |
| ``field:=json``,             | more fields need to be a ``Boolean``, ``Number``, |
| ``field:=@file.json``        | nested ``Object``, or an ``Array``,  e.g.,        |
|                              | ``meals:='["ham","spam"]'`` or ``pies:=[1,2,3]``  |
|                              | (note the quotes).                                |
+------------------------------+---------------------------------------------------+
| Fields upload fields         | Only available with ``--form, -f`` and            |
| ``field@/dir/file``          | ``--multipart``.                                  |
| ``field@file;type=mime``     | For example ``screenshot@~/Pictures/img.png``, or |
|                              | ``'cv@cv.txt;type=text/markdown'``.               |
|                              | With ``--form``, the presence of a file field     |
|                              | results in a ``--multipart`` request.             |
+------------------------------+---------------------------------------------------+


Note that the structured data fields aren’t the only way to specify request data:
The `raw request body`_ section describes mechanisms for passing arbitrary request data.


Escaping rules
--------------

You can use ``\`` to escape characters that shouldn’t be used as separators
(or parts thereof). For instance, ``foo\==bar`` will become a data key/value
pair (``foo=`` and ``bar``) instead of a URL parameter.

Often it is necessary to quote the values, e.g. ``foo='bar baz'``.

If any of the field names or headers starts with a minus
(e.g., ``-fieldname``), you need to place all such items after the special
token ``--`` to prevent confusion with ``--arguments``:

.. code-block:: bash

    $ http pie.dev/post  --  -name-starting-with-dash=foo -Unusual-Header:bar

.. code-block:: http

    POST /post HTTP/1.1
    -Unusual-Header: bar
    Content-Type: application/json

    {
        "-name-starting-with-dash": "foo"
    }



JSON
====

JSON is the *lingua franca* of modern web services and it is also the
**implicit content type** HTTPie uses by default.


Simple example:

.. code-block:: bash

    $ http PUT pie.dev/put name=John email=john@example.org

.. code-block:: http

    PUT / HTTP/1.1
    Accept: application/json, */*;q=0.5
    Accept-Encoding: gzip, deflate
    Content-Type: application/json
    Host: pie.dev

    {
        "name": "John",
        "email": "john@example.org"
    }


Default behaviour
-----------------


If your command includes some data `request items`_, they are serialized as a JSON
object by default. HTTPie also automatically sets the following headers,
both of which can be overwritten:

================    =======================================
``Content-Type``    ``application/json``
``Accept``          ``application/json, */*;q=0.5``
================    =======================================


Explicit JSON
-------------

You can use ``--json, -j`` to explicitly set ``Accept``
to ``application/json`` regardless of whether you are sending data
(it’s a shortcut for setting the header via the usual header notation:
``http url Accept:'application/json, */*;q=0.5'``). Additionally,
HTTPie will try to detect JSON responses even when the
``Content-Type`` is incorrectly ``text/plain`` or unknown.



Non-string JSON fields
----------------------

Non-string JSON fields use the ``:=`` separator, which allows you to embed arbitrary JSON data
into the resulting JSON object. Additionally, text and raw JSON files can also be embedded into
fields using ``=@`` and ``:=@``:

.. code-block:: bash

    $ http PUT pie.dev/put \
        name=John \                        # String (default)
        age:=29 \                          # Raw JSON — Number
        married:=false \                   # Raw JSON — Boolean
        hobbies:='["http", "pies"]' \      # Raw JSON — Array
        favorite:='{"tool": "HTTPie"}' \   # Raw JSON — Object
        bookmarks:=@files/data.json \      # Embed JSON file
        description=@files/text.txt        # Embed text file


.. code-block:: http

    PUT /person/1 HTTP/1.1
    Accept: application/json, */*;q=0.5
    Content-Type: application/json
    Host: pie.dev

    {
        "age": 29,
        "hobbies": [
            "http",
            "pies"
        ],
        "description": "John is a nice guy who likes pies.",
        "married": false,
        "name": "John",
        "favorite": {
            "tool": "HTTPie"
        },
        "bookmarks": {
            "HTTPie": "https://httpie.org",
        }
    }


Raw and complex JSON
--------------------

Please note that with the structured `request items`_ data field syntax, commands
can quickly become unwieldy when sending complex structures.
In such cases, it’s better to pass the full raw JSON data as a `raw request body`_, for example:

.. code-block:: bash

    $ echo -n '{"hello": "world"}' | http POST pie.dev/post

.. code-block:: bash

    $ http --raw '{"hello": "world"}' POST pie.dev/post

.. code-block:: bash

    $ http POST pie.dev/post < files/data.json

Furthermore, the structure syntax only allows you to send an object as the JSON document, but not an array, etc.
Here, again, the solution is to use a `raw request body`_.

Forms
=====

Submitting forms is very similar to sending `JSON`_ requests. Often the only
difference is in adding the ``--form, -f`` option, which ensures that
data fields are serialized as, and ``Content-Type`` is set to,
``application/x-www-form-urlencoded; charset=utf-8``. It is possible to make
form data the implicit content type instead of JSON
via the `config`_ file.


Regular forms
-------------

.. code-block:: bash

    $ http --form POST pie.dev/post name='John Smith'


.. code-block:: http

    POST /post HTTP/1.1
    Content-Type: application/x-www-form-urlencoded; charset=utf-8

    name=John+Smith


File upload forms
-----------------

If one or more file fields is present, the serialization and content type is
``multipart/form-data``:

.. code-block:: bash

    $ http -f POST pie.dev/post name='John Smith' cv@~/files/data.xml


The request above is the same as if the following HTML form were
submitted:

.. code-block:: html

    <form enctype="multipart/form-data" method="post" action="http://example.com/jobs">
        <input type="text" name="name" />
        <input type="file" name="cv" />
    </form>

Please note that ``@`` is used to simulate a file upload form field, whereas
``=@`` just embeds the file content as a regular text field value.

When uploading files, their content type is inferred from the file name. You can manually
override the inferred content type:

.. code-block:: bash

   $ http -f POST pie.dev/post name='John Smith' cv@'~/files/data.bin;type=application/pdf'

To perform a ``multipart/form-data`` request even without any files, use
``--multipart`` instead of ``--form``:

.. code-block:: bash

    $ http --multipart --offline example.org hello=world

.. code-block:: http

    POST / HTTP/1.1
    Content-Length: 129
    Content-Type: multipart/form-data; boundary=c31279ab254f40aeb06df32b433cbccb
    Host: example.org

    --c31279ab254f40aeb06df32b433cbccb
    Content-Disposition: form-data; name="hello"

    world
    --c31279ab254f40aeb06df32b433cbccb--

File uploads are always streamed to avoid memory issues with large files.

By default, HTTPie uses a random unique string as the multipart boundary
but you can use ``--boundary`` to specify a custom string instead:

.. code-block:: bash

    $ http --form --multipart --boundary=xoxo --offline example.org hello=world

.. code-block:: http

    POST / HTTP/1.1
    Content-Length: 129
    Content-Type: multipart/form-data; boundary=xoxo
    Host: example.org

    --xoxo
    Content-Disposition: form-data; name="hello"

    world
    --xoxo--

If you specify a custom ``Content-Type`` header without including the boundary
bit, HTTPie will add the boundary value (explicitly specified or auto-generated)
to the header automatically:


.. code-block:: bash

    http --form --multipart --offline example.org hello=world Content-Type:multipart/letter

.. code-block:: http

    POST / HTTP/1.1
    Content-Length: 129
    Content-Type: multipart/letter; boundary=c31279ab254f40aeb06df32b433cbccb
    Host: example.org

    --c31279ab254f40aeb06df32b433cbccb
    Content-Disposition: form-data; name="hello"

    world
    --c31279ab254f40aeb06df32b433cbccb--


HTTP headers
============

To set custom headers you can use the ``Header:Value`` notation:

.. code-block:: bash

    $ http pie.dev/headers  User-Agent:Bacon/1.0  'Cookie:valued-visitor=yes;foo=bar'  \

CHANGELOG.rst
==========
Change Log
==========

This document records all notable changes to `HTTPie <https://httpie.org>`_.
This project adheres to `Semantic Versioning <https://semver.org/>`_.



`2.5.0-dev`_ (unreleased)
-------------------------
* Added ``--raw`` to allow specifying the raw request body without extra processing as
  an alternative to ``stdin``. (`#534`_)
* Fixed ``--continue --download`` with a single byte to be downloaded left. (`#1032`_)
* Fixed ``--verbose`` HTTP 307 redirects with streamed request body. (`#1088`_)
* Add internal support for file-like object responses to improve adapter plugin support. (`#1094`_)
* Improved handling of responses without correct content-type charset. (`#1022`_)


`2.4.0`_ (2021-02-06)
---------------------
* Added support for ``--session`` cookie expiration based on ``Set-Cookie: max-age=<n>``. (`#1029`_)
* Show a ``--check-status`` warning with ``--quiet`` as well, not only when the output is redirected. (`#1026`_)
* Fixed upload with ``--session`` (`#1020`_).
* Fixed a missing blank line between request and response (`#1006`_).


`2.3.0`_ (2020-10-25)
-------------------------

* Added support for streamed uploads (`#201`_).
* Added support for multipart upload streaming (`#684`_).
* Added support for body-from-file upload streaming (``http pie.dev/post @file``).
* Added ``--chunked`` to enable chunked transfer encoding (`#753`_).
* Added ``--multipart`` to allow ``multipart/form-data`` encoding for non-file ``--form`` requests as well.
* Added support for preserving field order in multipart requests (`#903`_).
* Added ``--boundary`` to allow a custom boundary string for ``multipart/form-data`` requests.
* Added support for combining cookies specified on the CLI and in a session file (`#932`_).
* Added out of the box SOCKS support with no extra installation (`#904`_).
* Added ``--quiet, -q`` flag to enforce silent behaviour.
* Fixed the handling of invalid ``expires`` dates in ``Set-Cookie`` headers (`#963`_).
* Removed Tox testing entirely (`#943`_).


`2.2.0`_ (2020-06-18)
-------------------------

* Added support for custom content types for uploaded files (`#668`_).
* Added support for ``$XDG_CONFIG_HOME`` (`#920`_).
* Added support for ``Set-Cookie``-triggered cookie expiration (`#853`_).
* Added ``--format-options`` to allow disabling sorting, etc. (`#128`_)
* Added ``--sorted`` and ``--unsorted`` shortcuts for (un)setting all sorting-related ``--format-options``. (`#128`_)
* Added ``--ciphers`` to allow configuring OpenSSL ciphers (`#870`_).
* Added ``netrc`` support for auth plugins. Enabled for ``--auth-type=basic``
  and ``digest``, 3rd parties may opt in (`#718`_, `#719`_, `#852`_, `#934`_).
* Fixed built-in plugins-related circular imports (`#925`_).


`2.1.0`_ (2020-04-18)
---------------------

* Added ``--path-as-is`` to bypass dot segment (``/../`` or ``/./``)
  URL squashing (`#895`_).
* Changed the default ``Accept`` header value for JSON requests from
  ``application/json, */*`` to ``application/json, */*;q=0.5``
  to clearly indicate preference (`#488`_).
* Fixed ``--form`` file upload mixed with redirected ``stdin`` error handling
  (`#840`_).


`2.0.0`_ (2020-01-12)
-------------------------
* Removed Python 2.7 support (`EOL Jan 2020 <https://www.python.org/doc/sunset-python-2/>`_).
* Added ``--offline`` to allow building an HTTP request and printing it but not
  actually sending it over the network.
* Replaced the old collect-all-then-process handling of HTTP communication
  with one-by-one processing of each HTTP request or response as they become
  available. This means that you can see headers immediately,
  see what is being sent even if the request fails, etc.
* Removed automatic config file creation to avoid concurrency issues.
* Removed the default 30-second connection ``--timeout`` limit.
* Removed Python’s default limit of 100 response headers.
* Added ``--max-headers`` to allow setting the max header limit.
* Added ``--compress`` to allow request body compression.
* Added ``--ignore-netrc`` to allow bypassing credentials from ``.netrc``.
* Added ``https`` alias command with ``https://`` as the default scheme.
* Added ``$ALL_PROXY`` documentation.
* Added type annotations throughout the codebase.
* Added ``tests/`` to the PyPi package for the convenience of
  downstream package maintainers.
* Fixed an error when ``stdin`` was a closed fd.
* Improved ``--debug`` output formatting.


`1.0.3`_ (2019-08-26)
---------------------

* Fixed CVE-2019-10751 — the way the output filename is generated for
  ``--download`` requests without ``--output`` resulting in a redirect has
  been changed to only consider the initial URL as the base for the generated
  filename, and not the final one. This fixes a potential security issue under
  the following scenario:

  1. A ``--download`` request with no explicit ``--output`` is made (e.g.,
     ``$ http -d example.org/file.txt``), instructing httpie to
     `generate the output filename <https://httpie.org/doc#downloaded-filename>`_
     from the ``Content-Disposition`` response header, or from the URL if the header
     is not provided.
  2. The server handling the request has been modified by an attacker and
     instead of the expected response the URL returns a redirect to another
     URL, e.g., ``attacker.example.org/.bash_profile``, whose response does
     not provide  a ``Content-Disposition`` header (i.e., the base for the
     generated filename becomes ``.bash_profile`` instead of ``file.txt``).
  3. Your current directory doesn’t already contain ``.bash_profile``
     (i.e., no unique suffix is added to the generated filename).
  4. You don’t notice the potentially unexpected output filename
     as reported by httpie in the console output
     (e.g., ``Downloading 100.00 B to ".bash_profile"``).

  Reported by Raul Onitza and Giulio Comi.


`1.0.2`_ (2018-11-14)
-------------------------

* Fixed tests for installation with pyOpenSSL.


`1.0.1`_ (2018-11-14)

httpie/client.py
import argparse
import http.client
import json
import sys
from contextlib import contextmanager
from pathlib import Path
from typing import Callable, Iterable, Union
from urllib.parse import urlparse, urlunparse

import requests
# noinspection PyPackageRequirements
import urllib3
from . import __version__
from .cli.dicts import RequestHeadersDict
from .plugins.registry import plugin_manager
from .sessions import get_httpie_session
from .ssl import AVAILABLE_SSL_VERSION_ARG_MAPPING, HTTPieHTTPSAdapter
from .uploads import (
    compress_request, prepare_request_body,
    get_multipart_data_and_content_type,
)
from .utils import get_expired_cookies, repr_dict


urllib3.disable_warnings()

FORM_CONTENT_TYPE = 'application/x-www-form-urlencoded; charset=utf-8'
JSON_CONTENT_TYPE = 'application/json'
JSON_ACCEPT = f'{JSON_CONTENT_TYPE}, */*;q=0.5'
DEFAULT_UA = f'HTTPie/{__version__}'


def collect_messages(
    args: argparse.Namespace,
    config_dir: Path,
    request_body_read_callback: Callable[[bytes], None] = None,
) -> Iterable[Union[requests.PreparedRequest, requests.Response]]:
    httpie_session = None
    httpie_session_headers = None
    if args.session or args.session_read_only:
        httpie_session = get_httpie_session(
            config_dir=config_dir,
            session_name=args.session or args.session_read_only,
            host=args.headers.get('Host'),
            url=args.url,
        )
        httpie_session_headers = httpie_session.headers

    request_kwargs = make_request_kwargs(
        args=args,
        base_headers=httpie_session_headers,
        request_body_read_callback=request_body_read_callback
    )
    send_kwargs = make_send_kwargs(args)
    send_kwargs_mergeable_from_env = make_send_kwargs_mergeable_from_env(args)
    requests_session = build_requests_session(
        ssl_version=args.ssl_version,
        ciphers=args.ciphers,
        verify=bool(send_kwargs_mergeable_from_env['verify'])
    )

    if httpie_session:
        httpie_session.update_headers(request_kwargs['headers'])
        requests_session.cookies = httpie_session.cookies
        if args.auth_plugin:
            # Save auth from CLI to HTTPie session.
            httpie_session.auth = {
                'type': args.auth_plugin.auth_type,
                'raw_auth': args.auth_plugin.raw_auth,
            }
        elif httpie_session.auth:
            # Apply auth from HTTPie session
            request_kwargs['auth'] = httpie_session.auth

    if args.debug:
        # TODO: reflect the split between request and send kwargs.
        dump_request(request_kwargs)

    request = requests.Request(**request_kwargs)
    prepared_request = requests_session.prepare_request(request)
    if args.path_as_is:
        prepared_request.url = ensure_path_as_is(
            orig_url=args.url,
            prepped_url=prepared_request.url,
        )
    if args.compress and prepared_request.body:
        compress_request(
            request=prepared_request,
            always=args.compress > 1,
        )
    response_count = 0
    expired_cookies = []
    while prepared_request:
        yield prepared_request
        if not args.offline:
            send_kwargs_merged = requests_session.merge_environment_settings(
                url=prepared_request.url,
                **send_kwargs_mergeable_from_env,
            )
            with max_headers(args.max_headers):
                response = requests_session.send(
                    request=prepared_request,
                    **send_kwargs_merged,
                    **send_kwargs,
                )

            expired_cookies += get_expired_cookies(
                response.headers.get('Set-Cookie', '')
            )

            response_count += 1
            if response.next:
                if args.max_redirects and response_count == args.max_redirects:
                    raise requests.TooManyRedirects
                if args.follow:
                    prepared_request = response.next
                    if args.all:
                        yield response
                    continue
            yield response
        break

    if httpie_session and (
        httpie_session.is_new() or not args.session_read_only
    ):
        httpie_session.cookies = requests_session.cookies
        httpie_session.remove_cookies(
            # TODO: take path & domain into account?
            cookie['name'] for cookie in expired_cookies
        )
        httpie_session.save()


# noinspection PyProtectedMember
@contextmanager
def max_headers(limit):
    # <https://github.com/httpie/httpie/issues/802>
    # noinspection PyUnresolvedReferences
    orig = http.client._MAXHEADERS
    http.client._MAXHEADERS = limit or float('Inf')
    try:
        yield
    finally:
        http.client._MAXHEADERS = orig


def build_requests_session(
    verify: bool,
    ssl_version: str = None,
    ciphers: str = None,
) -> requests.Session:
    requests_session = requests.Session()

    # Install our adapter.
    https_adapter = HTTPieHTTPSAdapter(
        ciphers=ciphers,
        verify=verify,
        ssl_version=(
            AVAILABLE_SSL_VERSION_ARG_MAPPING[ssl_version]
            if ssl_version else None
        ),
    )
    requests_session.mount('https://', https_adapter)

    # Install adapters from plugins.
    for plugin_cls in plugin_manager.get_transport_plugins():
        transport_plugin = plugin_cls()
        requests_session.mount(
            prefix=transport_plugin.prefix,
            adapter=transport_plugin.get_adapter(),
        )

    return requests_session


def dump_request(kwargs: dict):
    sys.stderr.write(
        f'\n>>> requests.request(**{repr_dict(kwargs)})\n\n')


def finalize_headers(headers: RequestHeadersDict) -> RequestHeadersDict:
    final_headers = RequestHeadersDict()
    for name, value in headers.items():
        if value is not None:
            # “leading or trailing LWS MAY be removed without
            # changing the semantics of the field value”
            # <https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html>
            # Also, requests raises `InvalidHeader` for leading spaces.
            value = value.strip()
            if isinstance(value, str):
                # See <https://github.com/httpie/httpie/issues/212>
                value = value.encode('utf8')
        final_headers[name] = value
    return final_headers


def make_default_headers(args: argparse.Namespace) -> RequestHeadersDict:
    default_headers = RequestHeadersDict({
        'User-Agent': DEFAULT_UA
    })

    auto_json = args.data and not args.form
    if args.json or auto_json:
        default_headers['Accept'] = JSON_ACCEPT
        if args.json or (auto_json and args.data):
            default_headers['Content-Type'] = JSON_CONTENT_TYPE

    elif args.form and not args.files:
        # If sending files, `requests` will set
        # the `Content-Type` for us.
        default_headers['Content-Type'] = FORM_CONTENT_TYPE
    return default_headers


def make_send_kwargs(args: argparse.Namespace) -> dict:
    return {
        'timeout': args.timeout or None,
        'allow_redirects': False,
    }


def make_send_kwargs_mergeable_from_env(args: argparse.Namespace) -> dict:
    cert = None
    if args.cert:
        cert = args.cert
        if args.cert_key:
            cert = cert, args.cert_key
    return {
        'proxies': {p.key: p.value for p in args.proxy},
        'stream': True,
        'verify': {
            'yes': True,
            'true': True,
            'no': False,
            'false': False,
        }.get(args.verify.lower(), args.verify),
        'cert': cert,
    }


def make_request_kwargs(
    args: argparse.Namespace,
    base_headers: RequestHeadersDict = None,
    request_body_read_callback=lambda chunk: chunk
) -> dict:
    """
    Translate our `args` into `requests.Request` keyword arguments.

    """
    files = args.files
    # Serialize JSON data, if needed.
    data = args.data
    auto_json = data and not args.form
    if (args.json or auto_json) and isinstance(data, dict):
        data = json.dumps(data) if data else ''
    # Finalize headers.
    headers = make_default_headers(args)
    if base_headers:
        headers.update(base_headers)
    headers.update(args.headers)
    if args.offline and args.chunked and 'Transfer-Encoding' not in headers:
        # When online, we let requests set the header instead to be able more
        # easily verify chunking is taking place.
        headers['Transfer-Encoding'] = 'chunked'
    headers = finalize_headers(headers)

    if (args.form and files) or args.multipart:
        data, headers['Content-Type'] = get_multipart_data_and_content_type(
            data=args.multipart_data,
            boundary=args.boundary,
            content_type=args.headers.get('Content-Type'),
        )

    return {
        'method': args.method.lower(),
        'url': args.url,
        'headers': headers,
        'data': prepare_request_body(
            body=data,
            body_read_callback=request_body_read_callback,
            chunked=args.chunked,
            offline=args.offline,
            content_length_header_value=headers.get('Content-Length'),
        ),
        'auth': args.auth,
        'params': args.params.items(),
    }


def ensure_path_as_is(orig_url: str, prepped_url: str) -> str:
    """
    Handle `--path-as-is` by replacing the path component of the prepared
    URL with the path component from the original URL. Other parts stay
    untouched because other (welcome) processing on the URL might have
    taken place.

    <https://github.com/httpie/httpie/issues/895>


    <https://ec.haxx.se/http/http-basics#path-as-is>
    <https://curl.haxx.se/libcurl/c/CURLOPT_PATH_AS_IS.html>

    >>> ensure_path_as_is('http://foo/../', 'http://foo/?foo=bar')
    'http://foo/../?foo=bar'

    """
    parsed_orig, parsed_prepped = urlparse(orig_url), urlparse(prepped_url)
    final_dict = {
        # noinspection PyProtectedMember
        **parsed_prepped._asdict(),
        'path': parsed_orig.path,
    }
    return urlunparse(tuple(final_dict.values()))
